doc/developer/design/20320318_persist_columnar.md

```diff
@@ -0,0 +1,608 @@
+savings I don't believe there is a benefit to this approach. In fact it would
+incur some computational overhead to encode and there are no benefits from a
+SIMD perspective either.
+
+> **Alternative:** We could represent `Interval`s in a `StructArray` but we
+don't expose the internal details of `Interval` so this wouldn't aid in
+filtering or pushdown. The only benefit of structuring an interval would be for
+space reduction in dictionary encoding, but this is offset by the overhead of
+the struct itself.
+
+</td>
+
+<tr>
+<td> Jsonb </td>
+<td> 
+
+`TODO`
+
+</td>
+<td>
+
+`TODO`
+
+</td>
+<td>
+
+TODO.
+
+</td>
+
+<tr>
+<td> UUID </td>
+<td> 
+
+```rust
+extern crate uuid;
+
+uuid::Uuid([u8; 16])
+``` 
+
+</td>
+<td>
+
+`FixedSizeBinary[16]`
+
+</td>
+<td>
+
+Encode the bytes from the `Uuid` directly into a fixed size buffer.
+
+</td>
+
+<tr>
+<td> Array </td>
+<td> 
+
+```rust
+// DatumList<'a>.
+[T; N]
+``` 
+
+</td>
+<td>
+
+`FixedSizeList<T>[N]`
+
+</td>
+<td>
+
+A column of type array requires all of its elements to be the same size.
+
+> **Alternative:** We could use Arrow's Variable-size list, but it's less
+memory efficient
+
+</td>
+
+<tr>
+<td> List </td>
+<td> 
+
+```rust
+// DatumList<'a>.
+Vec<T>
+``` 
+
+</td>
+<td>
+
+`VariableSizeList<T>`
+
+</td>
+<td>
+
+Unlike `Array`s, `List`s are allowed to be ragged.
+
+</td>
+
+<tr>
+<td> Record </td>
+<td> 
+
+```rust
+Vec<(ColumnName, ColumnType)>
+``` 
+
+</td>
+<td>
+
+`TODO`
+
+</td>
+<td>
+
+TODO
+
+</td>
+
+<tr>
+<td> Map </td>
+<td> 
+
+```rust
+// DatumMap<'a>.
+HashMap<String, T>
+``` 
+
+</td>
+<td>
+
+`MapArray`
+
+</td>
+<td>
+
+The Arrow spec does not include the concept of a Map but the `arrow2` and
+`arrow-rs` crates have a `MapArray` type that is a list of tuples.
+
+> **Alternative:** We could encode maps to some binary format, e.g. proto, and
+store them as a binary blob. While this might be simpler it prevents us from
+being able to push down optimizations into the map.
+
+</td>
+
+<tr>
+<td> MzTimestamp </td>
+<td> 
+
+```rust
+struct MzTimestamp(u64);
+``` 
+
+</td>
+<td>
+
+`PrimitiveArray<u64>`
+
+</td>
+<td>
+
+Number of milliseconds since the UNIX epoch.
+
+</td>
+
+<tr>
+<td> Range </td>
+<td> 
+
+```rust
+struct Range<T> {
+  lower: RangeBound<T> {
+    inclusize: bool,
+    bound: Option<T>,
+  },
+  upper: RangeBound<T>,
+}
+``` 
+
+</td>
+<td>
+
+```
+RangeBound: StructArray<{
+  inclusive: bool,
+  bound: T,
+}>
+```
+
+```
+Range: StructArray<{
+  lower: RangeBound,
+  upper: RangeBound,
+}>
+```
+
+</td>
+<td>
+
+Structure the data as it is in Rust.
+
+Ranges seem pretty interesting and powerful, so Persist having an understanding
+of the data seems worthwhile for the long term. They could also be entirely
+unused (I'm not sure) in which case the complexity of encoding these in a
+structured way might not be worth it.
+
+> **Alternative:** Encode a Range into a binary format and store it as a blob.
```

Comment:
I'm definitely fine with the alternative for ranges if we want to use it as a place to save complexity/elbow grease.